tensorflow/python/ops/image_ops_impl.py

```diff
@@ -236,6 +236,66 @@ def _CheckAtLeast3DImage(image, require_static=True):
     return []
 
 
+def _AssertGrayscaleImage(image):
+  """Assert that we are working with a properly shaped
+     grayscale image.
+
+    Performs the check statically if possible (i.e. if the shape
+    is statically known). Otherwise adds a control dependency
+    to an assert op that checks the dynamic shape.
+
+    Args:
+      image: >= 3-D Tensor of size [*, height, width, depth]
+
+    Raises:
+      ValueError: if image.shape is not a [>= 3] vector or if 
+                last dimension is not size 1.
+
+    Returns:
+      If the shape of `image` could be verified statically, `image` is
+      returned unchanged, otherwise there will be a control dependency
+      added that asserts the correct dynamic shape.
+  """
+  return control_flow_ops.with_dependencies(
+      _CheckGrayscaleImage(image, require_static=False), image)
+
+
+def _CheckGrayscaleImage(image, require_static=True):
+    """Assert that we are working with properly shaped
+    grayscale image.
+
+    Args:
+      image: >= 3-D Tensor of size [*, height, width, depth]
+
+    Raises:
+      ValueError: if image.shape is not a [>= 3] vector or if 
+                last dimension is not size 1.
+
+    Returns:
+      An empty list, if `image` has fully defined dimensions. Otherwise, a list
+      containing an assert op is returned.
+    """
+    try:
+        if image.get_shape().ndims is None:
+            image_shape = image.get_shape().with_rank(3)
+        else:
+            image_shape = image.get_shape().with_rank_at_least(3)
+    except ValueError:
+        raise ValueError("A grayscale image must be at least three-dimensional.")
+    if require_static and not image_shape.is_fully_defined():
+      raise ValueError('\'image\' must be fully defined.')
+    if image_shape.is_fully_defined():
+        if image_shape[-1] != 1:
+            raise ValueError("Last dimension of a grayscale image should be size 1.")
+    if not image_shape.is_fully_defined():
```

Comment:
You should probably also assert dynamically that the rank is at least 3

Reply:
Isn't this already done in line 278-284?

Reply:
No, that verifies the static rank, not the result of tf.rank on the image

Reply:
Okay, my bad. I just tried to follow a similar structure as in `_CheckAtLeast3DImage`.